core/interfaces/i_selectable.js

```diff
@@ -0,0 +1,65 @@
+/**
+ * @license
+ * Copyright 2019 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @fileoverview The interface for an object that is selectable.
+ * @author samelh@google.com (Sam El-Husseini)
+ */
+
+'use strict';
+
+goog.provide('Blockly.ISelectable');
+
+goog.requireType('Blockly.WorkspaceSvg');
+
+
+/** @interface */
+Blockly.ISelectable = function() {};
+
+/**
+ * @type {string}
+ */
+Blockly.ISelectable.prototype.id;
+
+/**
+ * Get whether this is deletable or not.
+ * @return {boolean} True if deletable.
+ */
+Blockly.ISelectable.prototype.isDeletable;
+
+/**
+ * Get whether this is movable or not.
+ * @return {boolean} True if movable.
+ */
+Blockly.ISelectable.prototype.isMovable;
+
+/**
+ * Select this.  Highlight it visually.
+ * @return {void}
+ */
+Blockly.ISelectable.prototype.select;
+
+/**
+ * Unselect this.  Unhighlight it visually.
+ * @return {void}
+ */
+Blockly.ISelectable.prototype.unselect;
+
+/**
+ * Encode for copying.
+ * @return {!Blockly.ISelectable.CopyData} Copy metadata.
+ */
+Blockly.ISelectable.prototype.toCopyData;
+
+/**
+ * Copy Metadata.
+ * @typedef {{
+ *            xml:!Element,
+ *            source:Blockly.WorkspaceSvg,
+ *            typeCounts:Object
```

Comment:
`typeCounts` doesn't fit with the general copyable/selectable interface.  it's specific to blocks.

Reply:
Hmm. I can make it optional, but the ``copy_`` method expects typeCounts. (even if it is in some cases).
Not saying the copy / pasting logic doesn't need refactoring to make it more extensible.

Any ideas on what else can be done here?

Reply:
Yeah, I think it should have been optional there too--and it feels very strange to encode that into the defined interface.

But LGTM on this change, with a note to look at the copy/paste logic more generally.

Reply:
Sounds good. Filed https://github.com/google/blockly/issues/3833 to track.